Added SingleClassElementPerStatementFixer

phansys opened the pull request:
| Q             | A
| ------------- | ---
| Branch        | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1839
| License       | MIT
| Doc PR        |

According to PSR-2:
>There MUST NOT be more than one property declared per statement.

gharlan commented on src/Fixer/ClassNotation/SingleClassPropertyDeclarationFixer.php:
```diff
@@ -0,0 +1,172 @@
+ *
+ * @author Javier Spagnoletti <phansys@gmail.com>
+ */
+final class SingleClassPropertyDeclarationFixer extends AbstractFixer
```
We have the `SingleImportPerStatementFixer`. So I think this fixer should be named `SingleClassPropertyPerStatementFixer`.

phansys replied:
Updated fixer name.

gharlan commented:
Thanks for working on this!

I would like to see the same for class constants.

phansys commented:
What do you mean @gharlan? AFAIK this syntax isn't supported for class constants.

gharlan commented:
> What do you mean @gharlan? AFAIK this syntax isn't supported for class constants.

https://3v4l.org/jLJiC

gharlan commented on src/Fixer/ClassNotation/SingleClassPropertyDeclaredPerStatementFixer.php:
```diff
@@ -0,0 +1,172 @@
+ *
+ * @author Javier Spagnoletti <phansys@gmail.com>
+ */
+final class SingleClassPropertyDeclaredPerStatementFixer extends AbstractFixer
```
hmm, my proposal was `SingleClassPropertyPerStatementFixer`, without "Declared"

phansys replied:
Updated again.

SpacePossum commented on README.rst:
```diff
@@ -177,614 +177,631 @@ automatically fix anything:
-                         consecutive lines.
+                          Align double arrow symbols
+                          in consecutive lines.
 
```
what triggered the indent changes?

gharlan replied:
The longest fixer class name.. ;)

phansys commented:
@gharlan, yesterday when I've read your message I've tested that syntax in 3v4l.org too and I'm sure that failed (I don't know what can I've tested now).
BTW, that syntax isn't covered by PSR-2, so I guess it can be done later within a contrib fixer.

keradus commented:
@phansys same fixer, different configuration

phansys commented:
@keradus, it's possible to attach just a fixer configuration to a ruleset (PSR-2 in this case)?

keradus commented:
yes

keradus commented:
```
'@mySet' => {
    'fixer_A' => true, // no-settings / global settings
    'fixer_B' => [ 'some' => 'settings' ],
}
```

keradus commented:
any news here ?

phansys commented:
I'm very busy these days @keradus, my apologies. I'll try to find some time on weekend in order to continue this task. BTW, if somebody else wants to put some effort on it, I'll be very glad.

SpacePossum commented:
@phansys please see https://github.com/phansys/PHP-CS-Fixer/pull/1
let me know if you think it is helpful

With or without the PR this still needs work on the priorities.

phansys commented:
Thanks for the time you're dedicating to this @SpacePossum. I totally agree if you want to provide a PR in replacement of mine. Please let me know.

keradus commented:
@phansys if you able to finish this PR it would be great !

@SpacePossum thanks for your willingness for help, but please focus on currently opened PR of yours, like PHP7 bugfixes for 1.11 line ! I would say it's more important.

keradus commented:
btw, please regenerate the readme ;)

SpacePossum commented:
@phansys I have no time to take this one over, but I'm happy to help out or doing reviewing.
To wrap this one up I think a few details have to done:
- in my changes I insert the modifiers sequences without cloning the tokens, we need to make sure this is correct or change the code to do the cloning
- I can't think of priority issues with the fixer on prio. 0. However to be a bit more sure about it please run the fixer together with others on some projects (of your own and maybe something like Symfony, Twig, etc.). Run the fixer twice, the second time there shouldn't be any changes
- since it is PSR2 fixer it will be added to the Symfony by default, it would be nice to run the fixer on the Symfony project and if there are any hits please PR it @ Symfony, state it is the result of a new fixer and the goal of the PR is not to get it merged, but more of a heads up 

good luck!

SpacePossum commented:
btw. ping @gharlan I've added your sample case, can you confirm this is what you meant?
(https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1889/files#diff-35d2bcc89af69c2ad43f99869b477e02R46)

gharlan commented:
> btw. ping @gharlan I've added your sample case, can you confirm this is what you meant?

yes, thanks!

keradus commented on tests/Fixer/ClassNotation/SinglePropertyPerStatementFixerTest.php:
```diff
@@ -0,0 +1,406 @@
+use PhpCsFixer\Test\AbstractFixerTestCase;
+
+/**
+ * @author Javier Spagnoletti <phansys@gmail.com>
```
please make tests green.

phansys commented:
Test are green now @keradus.
Since this fixer takes care about class constants too, should we rename it? I propose "single_class_element_per_statement" or something similar.
I don't know if the same fixer could be used later as base for single variable declarations by instance.

SpacePossum commented:
fabbot is wrong, what happened to the bot?

keradus commented:
@phansys 
yes, please rename

@SpacePossum 
please, review and thumb if good to go

phansys commented:
I'll try to check this fixer today against some projects as @SpacePossum suggested.

phansys commented:
I've renamed the fixer, added some other tests and opened a [PR](https://github.com/symfony/symfony/issues/19176) against Symfony in order to get some feedback.
Ping @keradus, @SpacePossum.

GrahamCampbell commented on tests/Fixer/ClassNotation/SingleClassElementPerStatementFixerTest.php:
```diff
@@ -208,27 +208,21 @@ function doSomething()
 {
 
     public $bar = null;
-
```
?

phansys replied:
Which is your concern about this removed LF @GrahamCampbell?

GrahamCampbell replied:
Yes, I'm asking why this was removed.

phansys replied:
Because this fixer now uses [just the first previous line break](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1889/files#diff-fe9800e239d3b7335db6b921149d7a07R129) as separator, based on feedback from [Symfony](https://github.com/symfony/symfony/pull/19176#issuecomment-228517738).

SpacePossum commented on src/Fixer/ClassNotation/SingleClassElementPerStatementFixer.php:
```diff
@@ -0,0 +1,176 @@
+        $divisionContent = null;
+        if ($tokens[$startIndex - 1]->isWhitespace()) {
+            $divisionContent = $tokens[$startIndex - 1]->getContent();
+            if (preg_match('#(\n|\n\r)#', $divisionContent, $matches)) {
```
The sequence is typically `\r\n` IIRC, but it shouldn't matter, i.e. both cases should match.
So I think the match should be something like `(` 0-N times `\r` 1-N times `\n` 0-N times `\r` `)`.

keradus replied:
yes, `\r\n` for Windows

SpacePossum commented on src/Fixer/ClassNotation/SingleClassElementPerStatementFixer.php:
```diff
@@ -0,0 +1,176 @@
+            if ($divisionContent) {
+                $tokens->insertAt($i + 1, new Token(array(T_WHITESPACE, $divisionContent)));
+            }
+            $tokens->insertAt($i + 2, $modifiers);
```
@keradus I didn't clone the modifiers tokens here, but we seem to do this at other places.
Do we need it here, what is the reasoning/concept about it?

(missing empty line after `}` btw.)

keradus replied:
```php
$token = new Token([T_COMMENT, 'Foo']);

$tokens = Tokens::FromArray([$token, $token]);

$tokens[0]->getContent(); //Foo

$tokens[1]->setContent('Bar');
$tokens[1]->getContent(); //Bar

$tokens[0]->getContent(); //Bar :/ :/ :/
```

SpacePossum replied:
thanks for the sample code!
@phansys this means the modifiers inserted need to be cloned before inserted, could you update the PR?

keradus replied:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2024

keradus replied:
master is OK with this new check, this PR is not, checked ;)

SpacePossum replied:
tested a test, love the meta stuff of software development sometimes ;)

phansys replied:
Sure @SpacePossum, I'll try to do the required changes today. Thank you.

phansys replied:
Done.

keradus commented on src/Fixer/ClassNotation/SingleClassElementPerStatementFixer.php:
```diff
@@ -144,7 +142,9 @@ private function expandElement(Tokens $tokens, $startIndex, $endIndex)
             }
-            $tokens->insertAt($i + 2, $modifiers);
+            // collect modifiers
+            $sequence = $this->getModifiersSequences($tokens, $startIndex, $endIndex);
```
I'm not sure that calling `getModifiersSequences` method in loop is good, as it's already iterate through tokens in loop as well

keradus commented:
Could you squash consecutive commits of same author please?

phansys commented:
Of course @keradus, done.

keradus commented:
:+1:

SpacePossum commented:
:+1: 
LGTM

keradus commented:
Thank you @phansys.

keradus commented:
Thank you @phansys and @SpacePossum ! Great work !!!
